main.go

```diff
@@ -14,191 +14,336 @@ import (
-		},
-		[]string{"task", "mesos_slave", "framework_id"},
-	)
-	cpuUsrVal = prometheus.NewGaugeVec(
-		prometheus.GaugeOpts{
-			Namespace: "mesos_task",
-			Name:      "cpu_user_seconds_total",
-			Help:      "Cumulative user CPU time in seconds.",
-		},
-		[]string{"task", "mesos_slave", "framework_id"},
-	)
-	memLimitVal = prometheus.NewGaugeVec(
-		prometheus.GaugeOpts{
-			Namespace: "mesos_task",
-			Name:      "memory_limit_bytes",
-			Help:      "Task memory limit in bytes.",
-		},
-		[]string{"task", "mesos_slave", "framework_id"},
-	)
-	memRssVal = prometheus.NewGaugeVec(
-		prometheus.GaugeOpts{
-			Namespace: "mesos_task",
-			Name:      "memory_rss_bytes",
-			Help:      "Task memory RSS usage in bytes.",
-		},
-		[]string{"task", "mesos_slave", "framework_id"},
-	)
+	addr           = flag.String("web.listen-address", ":9105", "Address to listen on for web interface and telemetry")
+	autoDiscover   = flag.Bool("exporter.discovery", false, "Discover all Mesos slaves")
+	localAddr      = flag.String("exporter.local-address", "127.0.0.1", "Address to connect to the local Mesos slave")
+	masterURL      = flag.String("exporter.discovery.master", "http://mesos-master.example.com", "Mesos master URL")
+	metricsPath    = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics")
+	scrapeInterval = flag.Duration("exporter.interval", (60 * time.Second), "Scrape interval duration")
 )
 
-type Config struct {
-	Port      string      `json:"port"`
-	Interval  string      `json:"scrape_interval"`
-	Endpoints *[]Endpoint `json:"endpoints"`
+var httpClient = http.Client{
+	Timeout: 5 * time.Second,
 }
 
-type Endpoint struct {
-	Name       string        `json:"name"`
-	Target     string        `json:"target"`
-	Interval   string        `json:"scrape_interval,omitempty"`
-	TimeoutStr string        `json:"timeout,omitempty"`
-	Timeout    time.Duration `json:"-"`
+type ExporterOpts struct {
+	autoDiscover bool
+	interval     time.Duration
+	localAddr    string
+	masterURL    string
 }
 
-func newConfig() (*Config, error) {
-	var config Config
-
-	file, err := ioutil.ReadFile(*configFile)
-	if err != nil {
-		return nil, err
+type PeriodicExporter struct {
+	sync.RWMutex
+	errors  *prometheus.CounterVec
+	metrics []prometheus.Gauge
+	opts    *ExporterOpts
+	slaves  struct {
+		sync.Mutex
+		hostnames []string
 	}
-
-	err = json.Unmarshal(file, &config)
-	return &config, err
 }
 
-func request(e Endpoint) error {
-	client := http.Client{Timeout: 3 * time.Second}
-
-	resp, err := client.Get(e.Target + "/monitor/statistics.json")
-	if err != nil {
-		return err
+func NewMesosExporter(opts *ExporterOpts) *PeriodicExporter {
+	e := &PeriodicExporter{
+		errors: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: "exporter",
+				Name:      "scrape_errors",
+				Help:      "Current total scrape errors ",
+			},
+			[]string{"mesos_slave"},
+		),
+		opts:    opts,
+		metrics: make([]prometheus.Gauge, 0),
```

Comment:
No need to initialize empty slices (just omit this field from the initialization).